### ISSUE TITLE
Automated rancher/agent deployment

### ISSUE THREAD

**Issue by AnalogJ** [rancher/rancher#2599]
Hi,
I've started writing a chef cookbook to automate my setup of rancher and the associated agents on my servers. The issue I'm running into is that the agents all seem to need a unique registration token, which I cant seem to get/create using the api. 

Is there a way to do tokenless registration of rancher agents? Or retrieve unique registration tokens using the api?

**Comment by AnalogJ** [comment]
I was able to get it working by taking a closer look at the Vagrantfile. It seems you can ignore the token, but you still need to include base url.

`sudo docker run -d --privileged -v /var/run/docker.sock:/var/run/docker.sock rancher/agent:v0.8.2 http://192.168.0.x:8080`

This should probably be added to the docs at some point, but I was able to get everything working, so feel free to close the issue.

**Comment by vincent99** [comment]
They are unique to the installation but do not need to be unique to every host.  You generate one registration token (manually by going to the UI add custom host screen, or POST /v1/registrationTokens and reload the links.self URL until the state is active) and then can use that many times.

"Ignor[ing] tthe token" only works while access control is disabled.

**Comment by fentas** [comment]
My solution was to create an startup script, which will register hosts on startup.
Utilizing rancher rest api.

```sh
RANCHER_URL="..."
RANCHER_ACCESS_KEY="..."
RANCHER_SECRET_KEY="..."
## some host labels as needed
CATTLE_HOST_LABELS="host=$(hostname)&role=development"


## do some stuff like getting jq
sudo curl -skL https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64 -o /usr/local/bin/jq
sudo chmod +x /usr/local/bin/jq

## generate rancher registrationtokens
ID=$(curl -X POST \
-u "${RANCHER_ACCESS_KEY}:${RANCHER_SECRET_KEY}" \
-H 'Accept: application/json' \
-H 'Content-Type: application/json' \
-d "{\"name\":\"$(hostname)\"}" \
"${RANCHER_URL}/v1/registrationtokens" | jq -r .id)

## need to wait
sleep 5

COMMAND=$(curl -X GET \
-u "${RANCHER_ACCESS_KEY}:${RANCHER_SECRET_KEY}" \
-H 'Accept: application/json' \
"${RANCHER_URL}/v1/registrationtokens/$ID" | jq -r .command)

## start ranger-agent ~ running with root privileges no need for sudo. 
COMMAND=(docker run -e CATTLE_HOST_LABELS=$CATTLE_HOST_LABELS ${COMMAND#*run})
exec "${COMMAND[@]}"
```

Hope this helps.

**Comment by asciifaceman** [comment]
I created chef environments that match rancher environments for host/token segregation and managed the tokens via attributes

**Comment by asciifaceman** [comment]
eg.
`command "http://#{node['sce-rancher']['agent']['master']}/v1/scripts/#{node['sce-rancher']['agent']['tokens']}"`

**Comment by deniseschannon** [comment]
@AnalogJ As there have been a couple of responses for your request, I"m going to close the issue. Please open a new issue if you have a specific question.

**Comment by rcarmo** [comment]
Is there a way to set the token WHILE deploying the Rancher server? In that way, we could own that secret and bake it in to cloud-init scripts as machines are deployed.

**Comment by AnalogJ** [comment]
@rcarmo Are you trying to deploy the master server & agent nodes in parallel? It looks like you might have issues with that as the agents will fail if they can't talk to an already existing server. 

If you can run the agents after the server is up, then you can use the Rancher API to get a registrationToken:

https://github.com/mediadepot/chef-docker_rancher/blob/master/libraries/provider_agent.rb

**Comment by rcarmo** [comment]
@AnalogJ I know. I've done that in the past, but I think that still requires too much human intervention because you have to either have auth disabled or generate an API key. 

I _really_ want to be able to deploy everything in parallel, and it seems to me that being able to supply a shared secret to both masters and slaves would be much more practical - especially since on Azure I can instantiate a virtual machine scale set with that secret in CustomData and have new instances automatically register themselves - but I have to bake in the secret to the scale set definition.

So, in short, I would very much like to stop deploying Rancher clusters as a two-step process with manual intervention in between...

**Comment by AnalogJ** [comment]
@rcarmo Well generating the API key & enabling auth can be done in an automated way (https://github.com/mediadepot/chef-docker_rancher/blob/master/libraries/provider_auth_local.rb#L70-L109). But I can see where you're going with this. 

It's alot easier to just use a shared key than it is to 
- automatically generate the API key on master standup
- push the API key to a secret server/persistent storage/configuration management system
- use that API key for all subsequent agent installations

While a shared secret makes it easier, it is possible to have a completely automated cluster deployment.

**Comment by rcarmo** [comment]
Well, that works OK if your secret store is accessible from your master. If your secret store is not directly accessible from it, then you have to retrieve it to your orchestrator, save it and re-use it.

However, that too requires a second transaction for me, so it's a no-go.

**Comment by vincent99** [comment]
If auth is off then you don't need a registration token.  If it's on then you're already probably taking to the API to enable it. There is not a built in easy to inject a token for the default (or any other) project.  If your really want to you could just add an appropriate row to the database.

**Comment by rcarmo** [comment]
Hi @vincent99, I don't think you understood my goal at all. I need to be able to trigger creating and provisioning both master and worker VMs remotely, without human intervention, and supply them beforehand with all the required information they need to have a working cluster at the end.

At that time, there is no master (and certainly no database), and even if I create the master first, the orchestrator may not have access to its internals, including the database.

It would be _vastly_ easier if I could simply launch the the Rancher server container with a pre-defined environment variable (or even a mount point for a PEM certificate) and supply matching credentials to the workers.

Everything else (including the current approach) is just a kludge. Even `cloud-init` support would be nice, since I can supply instances with custom `cloud-init` files and inject anything I need into them beforehand.

**Comment by obriensystems** [comment]
Thanks guys, some details on this thread were very useful.
https://jira.onap.org/browse/OOM-715
https://gerrit.onap.org/r/#/c/32019/
https://jira.onap.org/browse/OOM-710
